### PR TITLE
Ben/scheduler

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/util/indico_estimator.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.framework import random_seed
```
Comment (benleetownsend) [inline finetune/util/indico_estimator.py]: Clean these up to use the public api

File: finetune/input_pipeline.py [hunk 2]
```diff
@@ -425,7 +425,7 @@ def get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None, contex
     def get_predict_input_fn(self, Xs, batch_size=None, context=None):
         batch_size = batch_size or self.config.predict_batch_size
         _, shapes = self.feed_shape_type_def()
-        tf_dataset = lambda: self._dataset_without_targets(Xs, train=None, context=context).padded_batch(batch_size, padded_shapes=shapes[0], drop_remainder=False)
+        tf_dataset = lambda: self._dataset_without_targets(Xs, train=None, context=context).padded_batch(batch_size, padded_shapes=shapes[0], drop_remainder=False).prefetch(tf.data.experimental.AUTOTUNE)
```
Comment (madisonmay) [inline finetune/input_pipeline.py]: Did this help?

File: finetune/scheduler.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import tensorflow as tf
+from finetune import Classifier
+
+class Scheduler:
+    def __init__(self, max_models=None):
+        self.loaded_models = list() # model queue
+        self.max_models = max_models
+        self.session = None
+        self.gpu_memory_limit = None
+        self.in_use_op = tf.contrib.memory_stats.BytesInUse()
```
Comment (madisonmay) [inline finetune/scheduler.py:10]: Seems like this could just be one op

File: finetune/scheduler.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        else:
+            out_model = self.model_cache[model]
+            self.loaded_models.remove(model) # put it back at the end of the queue
+            
+        self.loaded_models.append(model)
+        out_model._cached_predict = True
+        
+        return out_model
+
+    def predict(self, model_file, x):
```
Comment (madisonmay) [inline finetune/scheduler.py]: Should pass along any args / kwargs to the call to predict

File: finetune/scheduler.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    def _rotate_in_model(self, model):
+        if model not in self.loaded_models:
+            if (
+                    self.max_models is not None and len(self.loaded_models) + 1 > self.max_models_per_gpu or
+                    not self._memory_for_one_more()
+            ):
+                name = self.loaded_models.pop(0)
+                self.model_cache[name].close()
+                del self.model_cache[name]
+            out_model =	Classifier.load(model) # doesn't matter its classifier
```
Comment (madisonmay) [inline finetune/scheduler.py]: Some funky formatting

File: finetune/util/indico_estimator.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        self.predictions = None
+        self.mon_sess = None
+        self._cached_predict = False
+        super().__init__(*args, **kwargs)
+
+    def get_features_from_fn(self, input_fn, predict=True):
+        with tf.Graph().as_default() as g:
+            result = self._call_input_fn(input_fn, ModeKeys.PREDICT)
+            features, initializer = parse_input_fn_result(result)
+            if type(features) == tuple and predict:
```
Comment (madisonmay) [inline finetune/util/indico_estimator.py:30]: Worth a comment

File: finetune/util/indico_estimator.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                input_fn,
+                predict_keys=None,
+                hooks=None,
+                checkpoint_path=None,
+                yield_single_examples=True):
+        # Check that model has been trained.
+        self.g = self.g or tf.Graph()
+        random_seed.set_random_seed(self._config.tf_random_seed)
+        features_real, features = self.get_features_from_fn(input_fn)
+        with self.g.as_default():
```
Comment (madisonmay) [inline finetune/util/indico_estimator.py:63]: Could probably break up this fn but low priority

File: finetune/base.py [hunk 8]
```diff
@@ -322,6 +325,7 @@ def _get_estimator_config(self):
             optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1 
 
         distribute_strategy = self._distribute_strategy(self.config.visible_gpus)
+        print(self.config.val_interval)
```
Comment (madisonmay) [inline finetune/base.py]: rogue print

**Review by madisonmay — APPROVED** [review]

**Pull request merged**